sql/src/main/java/org/apache/druid/sql/calcite/planner/DruidSqlValidator.java

```diff
@@ -475,18 +476,27 @@ private RelDataType validateTargetType(
         continue;
       }
       SqlTypeName sqlTypeName = SqlTypeName.get(definedCol.sqlStorageType());
-      RelDataType relType = typeFactory.createSqlType(sqlTypeName);
-      if (NullHandling.replaceWithDefault() && !SqlTypeFamily.STRING.contains(relType)) {
-        fields.add(Pair.of(
-            colName,
-            relType
-        ));
+      RelDataType relType;
+      if (sqlTypeName != null) {
+        relType = typeFactory.createSqlType(sqlTypeName);
       } else {
-        fields.add(Pair.of(
-            colName,
-            typeFactory.createTypeWithNullability(relType, true)
-        ));
+        ColumnType columnType = ColumnType.fromString(definedCol.sqlStorageType());
+        if (columnType != null && columnType.getType().equals(ValueType.COMPLEX)) {
+          relType = RowSignatures.makeComplexType(typeFactory, columnType, sourceField.getType().isNullable());
+        } else {
+          relType = RowSignatures.columnTypeToRelDataType(
+              typeFactory,
+              columnType,
+              // this nullability is ignored for complex types for some reason, hence the check for complex above.
+              sourceField.getType().isNullable()
+          );
+        }
       }
```

Comment:
this seems to me that it tries to compute the `RelDataType` for `definedCol` ; can we have method or something for that? maybe that could even make it easier to write more direct tests

Reply:
Good call, fixed